### PR TITLE
[DO-NOT-REVIEW/WIP] Remove datacenters.yaml from controllers

### PR DESCRIPTION
**Pull request by mrIncompetent**
**What this PR does / why we need it**:
To make the whole controller work more independent and self-contained the datacenters.yaml is being removed.

Whats broken:
- Updating a cluster via the UI
  Reason here is that the UI does not send the newly added fields as those fields need to be implemented in the dashboard before. I will create a follow-up issue after merging.
- Defaulting of the vSphere vmFolder

Changes to be made:
- Move defaulting into a Middleware & introduce middleware only for new Project-related create-cluster endpoint
  We're going to change the create-cluster endpoint to make the whole cluster struct part of the Request body - instead of just the ClusterSpec as we now have. This will make the endpoint REST conform and enable a proper defaulting middleware (We need things like the cluster-name inside the defaulting)

### CODE REVIEW

**Comment by glower** [comment]
It's looking good to me, but someone else need to have a look, it is huge!

File: api/pkg/crd/kubermatic/v1/cluster.go [hunk 1]
```diff
@@ -207,13 +207,21 @@ type AzureCloudSpec struct {
...
+	Location       string `json:"location"`
 }
 
 // VSphereCloudSpec specifies access data to VSphere cloud.
 type VSphereCloudSpec struct {
 	Username  string `json:"username"`
 	Password  string `json:"password"`
 	VMNetName string `json:"vmNetName"`
+
+	Endpoint      string `json:"endpoint"`
```
Comment (alvaroaleman) [inline api/pkg/crd/kubermatic/v1/cluster.go:219]: Now that we have this combined in one place anyways, can we maybe just import the `config` types for the cloud provides from the machine-controller? E.g.  https://github.com/kubermatic/machine-controller/blob/master/pkg/cloudprovider/provider/aws/provider.go#L149-L166

File: api/pkg/datacenter/merge.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	if dc.Spec.VSphere == nil {
+		return nil, errors.New("node datacenter has no vSphere specification")
+	}
+
+	if spec.VSphere.Endpoint == "" {
+		spec.VSphere.Endpoint = dc.Spec.VSphere.Endpoint
+		spec.VSphere.AllowInsecure = dc.Spec.VSphere.AllowInsecure
+	}
+
+	if spec.VSphere.Cluster == "" {
```
Comment (alvaroaleman) [inline api/pkg/datacenter/merge.go:61]: `AllowInsecure` is missing

File: api/pkg/datacenter/merge.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+	return spec, nil
+}
+
+func defaultFromAWSDatacenter(spec *kubermaticv1.CloudSpec, dc provider.DatacenterMeta) (*kubermaticv1.CloudSpec, error) {
+	if dc.Spec.AWS == nil {
+		return nil, errors.New("node datacenter has no AWS specification")
+	}
+
+	if spec.AWS.Region == "" {
```
Comment (alvaroaleman) [inline api/pkg/datacenter/merge.go:87]: `Zone` is missing

File: api/pkg/datacenter/merge.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+// DefaultFromDatacenter adds defaults coming from the datacenter to the cluster object
+func DefaultFromDatacenter(spec *kubermaticv1.CloudSpec, dc provider.DatacenterMeta) (*kubermaticv1.CloudSpec, error) {
+	if spec.Digitalocean != nil {
+		return spec, nil
+	} else if spec.Openstack != nil {
+		return defaultFromOpenstackDatacenter(spec, dc)
+	} else if spec.VSphere != nil {
+		return defaultFromVSphereDatacenter(spec, dc)
+	} else if spec.Azure != nil {
```
Comment (alvaroaleman) [inline api/pkg/datacenter/merge.go:19]: What about `AzureCloudSpec.Location`?

**Pull request closed without merging**